### ISSUE TITLE
Concurrent PUTs resulting in hfj_res_ver.IDX_RESVER_ID_VER violations - locking options? 

### ISSUE THREAD

**Issue by XcrigX** [hapifhir/hapi-fhir#4690]
We are sending Transaction bundles in parallel to the server with PUT operations that may contain the same resources such as Organization, Patient, Practitioner, etc. 

We are periodically seeing UNIQUE KEY constraint violations for hfj_res_ver.IDX_RESVER_ID_VER when processing the same resource in parallel.  I surmise multiple threads are looking up the max version of the resource in parallel, then both trying to insert the same next version. 

I had thought to add an interceptor that uses a striped locking mechanism to prevent the same resource from being processed in parallel on multiple threads. 
But thought I'd raise it here first. Are there any existing options to prevent this? Any foreseeable issues with adding something to do it? 

simplistically, something like this: 

```
//see https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html
private Striped<Lock> resourceLockMgr = Striped.lock(64); //64 is arbitrary, the number of buckets
...
String lockKey = res.getIdElement().getResourceType() + "/" + res.getIdElement().getIdPart();
//wait for the lock before proceeding
Lock resourceLock = resourceLockMgr.get(lockKey);
resourceLock.lock();
try{
	//commit the resource
}
finally {
	resourceLock.unlock();
}
```

Thoughts?

**Comment by G-2-Z** [comment]
We have a similar issue updating resources quickly using the DAO, but even in the same thread. https://github.com/hapifhir/hapi-fhir/issues/4843

**Comment by jamesagnew** [comment]
Closing as duplicate of https://github.com/hapifhir/hapi-fhir/issues/4843 (which I realized after the fact is newer than this ticket.... but it certainly looks like the same issue)
